### ISSUE TITLE
IPV6 Support is Required for App Store Submission

### ISSUE THREAD

**Issue by svarlamov** [spotify/ios-sdk#676]
Since June 1st, the App Store is now requiring that all Spotify apps support IPV6. Our recent submission was rejected, because they weren't able to play any tracks, and they suggested that it was likely due to the fact that the Spotify iOS SDK may not be compatible. All other parts of the app were entirely functional, and the app had no issue with calling our own servers that are properly setup for IPV6 connections. Is it possible that the iOS SDK isn't ready for this new requirement from Apple, and if so, how could we solve it? Thanks for the help, as we would really like to deliver our app to the public as soon as possible

**Comment by volksport** [comment]
I can't +1 this hard enough.

**Comment by svarlamov** [comment]
We have since verified that the lack of IPv6 support in the SDK was the cause of our rejection, as we replicated the issue when testing our app on an IPv6 only network. Would really be wonderful to see some support for this critical issue, or at least some thoughts on when we could possibly see a fix from @spotify

**Comment by ahallora** [comment]
This +1!!

**Comment by erdtman** [comment]
We have seen this and are working on a solution, but it is not a quick fix so it could take a few weeks.

We are relay sorry about the inconvenience that this creates.

**Comment by timbodeit** [comment]
As far as I can see the only place where Apple announced the IPv6 requirement was a news article named ["Supporting IPv6-only Networks"](https://developer.apple.com/news/?id=05042016a).

Am I the only one who finds it amusing, that this article can not be accessed from an IPv6-only Network? 😝

**Comment by CVertex** [comment]
We're waiting on this too! Please keep us informed when you have a timeline in mind

**Comment by ddramowicz** [comment]
+1! Thanks a lot.

**Comment by loretoparisi** [comment]
+1 that's a pretty huge requirement. I wonder at which level the IPv6 is missing since this framework should use the [SPTDataLoader](https://github.com/spotify/SPTDataLoader/blob/master/SPTDataLoader/SPTDataLoaderRequest.m) library that uses a `NSMutableURLRequest` after all:

```objective-c
- (NSURLRequest *)urlRequest
{
    NSString * const SPTDataLoaderRequestContentLengthHeader = @"Content-Length";
    NSString * const SPTDataLoaderRequestAcceptLanguageHeader = @"Accept-Language";
    
    NSMutableURLRequest *urlRequest = [NSMutableURLRequest requestWithURL:self.URL];
    
    if (!self.headers[SPTDataLoaderRequestAcceptLanguageHeader]) {
        [urlRequest addValue:[self.class languageHeaderValue]
          forHTTPHeaderField:SPTDataLoaderRequestAcceptLanguageHeader];
    }
    
    if (self.body) {
        [urlRequest addValue:@(self.body.length).stringValue forHTTPHeaderField:SPTDataLoaderRequestContentLengthHeader];
        urlRequest.HTTPBody = self.body;
    }
    
    NSDictionary *headers = self.headers;
    for (NSString *key in headers) {
        NSString *value = headers[key];
        [urlRequest addValue:value forHTTPHeaderField:key];
    }
    
    urlRequest.cachePolicy = self.cachePolicy;
    urlRequest.HTTPMethod = NSStringFromSPTDataLoaderRequestMethod(self.method);
    
    return urlRequest;
}
```

**Comment by timbodeit** [comment]
@erdtman Now that two weeks have passed, can you give us a more detailed estimate when we can expect a fix?

**Comment by bsatorius** [comment]
Does Apple's new policy apply to both new submissions and updates or just new apps only?

**Comment by timbodeit** [comment]
@bsatorius As far as I know, it applies to both new submissions and updates.

**Comment by bsatorius** [comment]
@timbodeit thanks, was hoping to get approval on an update today, so looks like we may end up in a holding pattern.

**Comment by erdtman** [comment]
@timbodeit we expect it to take a 2-3 weeks more before we have release with full IPv6 support.

We are really sorry for the inconveniences that this causes.

**Comment by CVertex** [comment]
@erdtman @timbodeit D'oh! That's okay. Thanks for the update.. Good luck with the upgrade. Let us know how we can help ;)

**Comment by Muesly** [comment]
+1

**Comment by erdtman** [comment]
A few weeks has passed and it is time for an update.
We are making progress and the big part of the work is done. We now have to quality assure the release before it goes out, i.e. expect it to be a few weeks more before it is publicly available.

Sorry for the inconveniences this causes.

**Comment by timbodeit** [comment]
How complicated can a patch to add IPv6 support be, that it needs multiple weeks of quality assurance? 😳

I do understand, that a server-side IPv6 rollout would take time. But a client-side framework?

**Comment by CVertex** [comment]
@erdtman Thanks for the update. Hope testing goes smoothly

@timbodeit They're probably using some C++ library under the hood that is tough to maintain

**Comment by swillsea** [comment]
Thanks for the update @erdtman. Looking forward to getting the latest version soon.

**Comment by NathanWalker** [comment]
I too am curious every day to hear updates on this. I have been working on an app for months with this spotify ios sdk and just submitted to the store last week to learn that it was rejected to it not supporting IPv6. Beyond disappointed. Months of work :( Was planning on giving a talk over the app development next month with it being available on the store. You guys have done tremendous work on this sdk and I have thoroughly enjoyed using it with my premium account and would love to help this effort in any way, please let any of us know if we can help make this a reality, like tomorrow 👍

**Comment by jlooper** [comment]
This is a big +1 from me as well!

**Comment by philiparpin** [comment]
Any updates? You guys seem to be running severely behind schedule on this, which is beginning to impact our development schedules and hinder our progress.

**Comment by thanasistratigakis** [comment]
My team has re-scheduled our app launch three times because of this issue. Any updates on full IPv6 support?

**Comment by NathanWalker** [comment]
At minimum, a quick comment, even if all it says is:'
"We have encountered significant roadblocks that..."
"At the moment we are held up by..."
"We are as anxious as all of you are to solve this but..."

No response causes more panic as it is easily assumed your team may have no idea what to do due to unforeseen issues.

**Comment by ddramowicz** [comment]
Yes, please let us know what the status is! We'd appreciate it a lot. I am also happy to help in any way!

**Comment by brianhans** [comment]
Any updates on when the changes will be pushed to github? This is holding me back from pushing my app to the store.

**Comment by d-kunin** [comment]
We are really sorry to make you wait and block all the awesome releases. We are working on the major update of the underlying implementation of the iOS SDK and IPv6 part is already fixed, but there are few issues that we still need to fix to provide you with functional solution. The earliest possible release date is the middle of next week. There will be API changes as well, so be ready to adapt your implementation. I provide the changelist below, but it might be changed as we approach the release:

```
Updated functions
-(void)playURI:(NSURL *)uri callback:(SPTErrorableOperationCallback)block;
-(void)queueURI:(NSURL *)uri callback:(SPTErrorableOperationCallback)block;

Most of the changes made has been announced by deprecation of functions in previous releases, but some of the changes has not.

Functions not deprecated that has been removed
-(void)playURIs:(NSArray *)uris fromIndex:(int)index callback:(SPTErrorableOperationCallback)block;
-(void)playURIs:(NSArray *)uris withOptions:(SPTPlayOptions *)options callback:(SPTErrorableOperationCallback)block;
-(void)replaceURIs:(NSArray *)uris withCurrentTrack:(int)index callback:(SPTErrorableOperationCallback)block;
-(void)queueURIs:(NSArray *)uris clearQueue:(BOOL)clear callback:(SPTErrorableOperationCallback)block;
@property (nonatomic, readonly, assign) BOOL initialized;
@property (nonatomic, readonly) int trackListSize;

Deprecated functions that has been removed
-(void)playURI:(NSURL *)uri fromIndex:(int)index callback:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)setURIs:(NSArray *)uris callback:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)playURIsFromIndex:(int)index callback:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)playTrackProvider:(id<SPTTrackProvider>)provider callback:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)playTrackProvider:(id<SPTTrackProvider>)provider fromIndex:(int)index callback:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)queueURI:(NSURL *)uri clearQueue:(BOOL)clear callback:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)queueTrackProvider:(id<SPTTrackProvider>)provider clearQueue:(BOOL)clear callback:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)queuePlay:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)queueClear:(SPTErrorableOperationCallback)block DEPRECATED_ATTRIBUTE;
-(void)getRelativeTrackMetadata:(int)index callback:(void (^)(NSDictionary *))block DEPRECATED_ATTRIBUTE;
-(void)getAbsoluteTrackMetadata:(int)index callback:(void (^)(NSDictionary *))block DEPRECATED_ATTRIBUTE;
@property (nonatomic, readwrite) int trackListPosition DEPRECATED_ATTRIBUTE;
@property (nonatomic, readonly) int queueSize DEPRECATED_ATTRIBUTE;
```
We are working hard to provide you with the update, once again sorry for inconvenience.

**Comment by mhuusko5** [comment]
Personally I don't believe the above API changes are going to affect me – I don't use the SDK's queueing functionality/player metadata calls etc.  – but I _believe_ (someone feel free to correct me) it's universal bad practice to in a single release both make sweeping API changes as well as fix a show/release-stopping bug like the IPv6 bug.

This is of course for the simple reason that it means for users to adopt the new beta/IPv6 fix (read "start shipping their app again"!) they also need to pay the cost of working around these API changes, which depending on the user/app/use of above related APIs, could be significant work... idk.. we'll see when others weigh in.

And lets remember, many of us are on betas even before the current one (14, 15, etc.) to avoid other show stopping bugs (https://github.com/spotify/ios-sdk/issues/650, https://github.com/spotify/ios-sdk/issues/697 , etc.) introduced/present in 16/17. So, even if you fix this IPv6 bug in beta 18 so its physically possible to submit apps to the store, we still won't be able to adopt the beta else we'll be shipping these other bugs making our apps near unusable.

TL;DR: We need a beta 18 with IPv6 fixed, no new bugs, and the major bugs introduced in 16/17 fixed, so we can upgrade from 15 and ship our apps.

**Comment by CVertex** [comment]
I'm in agreement with @mhuusko5 that it's a bit drastic to remove APIs when you only needed to update for IPv6.
The level of communication in this SDK is astoundingly poor. 

We use the queuing APIs and this changes lots of our code. 
Why have functions gone straight from active to deleted without going via deprecation? 

Perhaps some elaboration on the rationale for these changes would be positive for everyone in this community.

**Comment by NathanWalker** [comment]
For me personally, I don't use the queue apis and based on your changeset, I would not be affected at all with my app, so I'm fine with your current course. Bring on the IPv6 :) Thanks for the effort in bringing this support that many of us are eagerly awaiting.

My app is currently using beta 17 with no issues (aside from IPv6 support) so looking forward to beta18!

**Comment by mhuusko5** [comment]
re: "no issues", @NathanWalker ever tried switching to LTE? 😜

(below not directed at you @NathanWalker btw. maybe you've just been lucky 😄. cherish it!)

But seriously, let's not pretend there's not 240 open issues, 50 or so non-dupe/valid/bugs, a dozen or so show-stopper level, and a handful made worse or more difficult to work around since beta 16/17. We've got a community of people using this SDK to look after here, some of which using it at a very deep and business critical level.

Same as if you have an operating system with multiple active releases/people stuck on different versions for one reason or another, when you have a critical (say, security) fix you need everyone to adopt, it would be appropriate to apply that fix to multiple versions.
As it's unlikely beta 18 can suddenly have everything from after 14/15 fixed along with the new IPv6 fix, an alternative solution would be to apply the underlying IPv6 fix to 14/15 and re-release those versions otherwise unchanged.

**Comment by d-kunin** [comment]
@CVertex @mhuusko5 we would be happy to introduce "just IPv6 fix", but it was technically not possible since, as I have mentioned, internal implementation has been changed drastically.
Unfortunately, it is not always possible to follow best practices. These changes have to be made to move the SDK from old and mostly unmaintained underlying 'core' C/C++ parts to the most recent version of 'core' that is being actively developed.

**Comment by mhuusko5** [comment]
@d-kunin Okay, understood. Is there any chance Beta 18, as part of adopting this actively developed 'core', will include fixes along with the IPv6 fix, that will make it adoptable for those of us still on Beta 14/15 (e.g. improvements for LTE switch, fixes for via-app login failing, or re-addition of callback based player-login features for working around the LTE switch problem)?

**Comment by NOlivNeto** [comment]
I understand, you've done a major refactor on the sdk and it's hard to do everything without breaking stuff... But could you please try to deprecate a method before removing it next time? I thank you for saying it early, so I can start right away, but, please, if a release is so urgent like this, we have a very bad time fixing everything as soon as the new beta is released.

**Comment by twof** [comment]
>The earliest possible release date is the middle of next week.
Any updates? We were hoping to ship tonight or tomorrow.

**Comment by gchilds** [comment]
Please, we're dying out here.

**Comment by MuruMusic** [comment]
Spotify crew -  Whats the ETA on this? Would be super useful to know,  not to mention the courteous thing to do.

**Comment by CVertex** [comment]
What's the ETA on this? Would be super useful to know... not to mention the courteous thing to do.

**Comment by NOlivNeto** [comment]
`The earliest possible release date is the middle of next week.`
It's been 14 days now. When you give a date, you create expectation and project planning... If you don't have a reliable date, please don't give any.

**Comment by d-kunin** [comment]
We are starting to prepare the release tomorrow, so if everything goes as planned, it will be shipped in a few days. Fingers crossed.

**Comment by d-kunin** [comment]
Up and running! Please check the changelog for more info: https://github.com/spotify/ios-sdk/blob/master/CHANGELOG.md

**Comment by NathanWalker** [comment]
@d-kunin the changelog mentions:
```
API Changes Updated functions:

-(void)playURI:(NSURL *)uri callback:(SPTErrorableOperationCallback)block;
```
But what was the update/change?  I can no longer get that function to work. It will execute the call but the callback block never gets called? What needs to be updated with the code to make it work?
